src/MongoDB/Manager.c

```diff
@@ -254,20 +254,33 @@ static void php_phongo_manager_prep_uri_options(zval* options TSRMLS_DC) /* {{{
 /* Selects a server for an execute method. If "for_writes" is true, a primary
  * will be selected. Otherwise, a read preference will be used to select the
  * server. If zreadPreference is NULL, the client's read preference will be
- * used.
+ * used. If zsession is a session object in a sharded transaction, the session
+ * will be checked whether it is pinned to a server. If so, that server will be
+ * selected. Otherwise, server selection
  *
  * On success, server_id will be set and the function will return true;
  * otherwise, false is returned and an exception is thrown. */
-static bool php_phongo_manager_select_server(bool for_writes, zval* zreadPreference, mongoc_client_t* client, uint32_t* server_id TSRMLS_DC) /* {{{ */
+static bool php_phongo_manager_select_server(bool for_writes, zval* zreadPreference, zval* zsession, mongoc_client_t* client, uint32_t* server_id TSRMLS_DC) /* {{{ */
 {
-	const mongoc_read_prefs_t*   read_preference = NULL;
 	mongoc_server_description_t* selected_server;
-	bson_error_t                 error = { 0 };
+	const mongoc_read_prefs_t*   read_preference = NULL;
+	bson_error_t                 error           = { 0 };
 
 	if (!for_writes) {
 		read_preference = zreadPreference ? phongo_read_preference_from_zval(zreadPreference TSRMLS_CC) : mongoc_client_get_read_prefs(client);
 	}
 
+	if (zsession) {
```

Comment:
Do you think this should pre-empty the `if (!for_writes)` condition above? If we end up using a pinned server, there's no need to fetch `read_preference`. Swapping these would keep that related code closer together.

Reply:
You're absolutely correct. I've switched the order to avoid parsing the read preference when we're using a pinned server anyways.